Dead letter queue support

nataliezeller1 opened the pull request:
When DLQ is configured, move malformed records (ones that fail to be indexed in OpenSearch due to mapping issues) to the DLQ topic.
Records moved to DLQ will be added with a header named "__connect.errors.exception.message", containing the reason for the failure (the error message as reported by OpenSearch).

C0urante commented on src/main/java/io/aiven/kafka/connect/opensearch/BulkProcessor.java:
```diff
@@ -401,6 +421,25 @@ private BulkResponse execute() throws Exception {
+                    LOGGER.info("Reporting on record failure to DLQ");
+                    final SinkRecord record = batch.get(itemResponse.getItemId()).getSinkRecord();
+                    if (!(record.value() instanceof Map)) {
+                        LOGGER.error("Only Map objects supported");
```
Why do we have this check here? Doesn't the connector support arrays and structs as well?

nataliezeller1 replied:
This was here by mistake, removed it.

C0urante reviewed (commented):
Thank @nataliezeller1! Great to see this feature being added to the connector. I've left a few comments; let me know what you think. Have not reviewed most of the test portions yet; if any of my remarks can be addressed with testing bits, feel free to direct me to them and I'll read up.

nataliezeller1 commented on src/integration-test/java/io/aiven/kafka/connect/opensearch/OpensearchSinkTaskIT.java:
```diff
@@ -130,6 +132,8 @@ public void testCompatible() throws Exception {
 
         final var opensearchSinkTask = new OpensearchSinkTask();
         try {
+            final var mockContext = mock(WorkerSinkTaskContext.class);
```
Fixed

nataliezeller1 commented on src/integration-test/java/io/aiven/kafka/connect/opensearch/OpensearchSinkTaskIT.java:
```diff
@@ -350,6 +354,8 @@ protected Struct createOtherRecord(final Schema schema) {
 
     private void runTask(final Map<String, String> props, final List<SinkRecord> records) {
         final var opensearchSinkTask = new OpensearchSinkTask();
+        final var mockContext = mock(WorkerSinkTaskContext.class);
```
Fixed

nataliezeller1 commented on src/main/java/io/aiven/kafka/connect/opensearch/OpensearchSinkTask.java:
```diff
@@ -74,7 +74,7 @@ public void start(final Map<String, String> props) {
             }
 
-            this.client = new OpensearchClient(config);
+            this.client = new OpensearchClient(config, context.errantRecordReporter());
```
Fixed

nataliezeller1 commented on src/main/java/io/aiven/kafka/connect/opensearch/BulkProcessor.java:
```diff
@@ -457,7 +496,27 @@ private void handleMalformedDoc(final BulkItemResponse bulkItemResponse) {
+            this.sinkRecord = sinkRecord;
+        }
+
+        public DocWriteRequest<?> getDocWriteRequests() {
```
Fixed

nataliezeller1 commented on src/main/java/io/aiven/kafka/connect/opensearch/BulkProcessor.java:
```diff
@@ -457,7 +496,27 @@ private void handleMalformedDoc(final BulkItemResponse bulkItemResponse) {
     }
-    
+
+    public final class DocWriteRequestWrapper {
```
Fixed

nataliezeller1 commented on src/main/java/io/aiven/kafka/connect/opensearch/BulkProcessor.java:
```diff
@@ -367,14 +383,18 @@ private BulkResponse execute() throws Exception {
                     }
                     for (final var itemResponse : response.getItems()) {
                         if (itemResponse.isFailed()) {
+                            reportMalformedRecord(itemResponse);
```
Changed the logic to report to DLQ only after all retry attempts are completed. 
This will happen only in case DLQ is configured (otherwise reporter will be null and reportMalformedRecords method will return with doing nothing).
Note that in case of an "aborted" failure, the records will be written to DLQ but the exception will be thrown as before, making the connector to stop.
In case a response was returned from OpenSearch, the error reported to DLQ will contain the rest status, action id and error message (as suggested). In case an exception was thrown (no response from Opensearch), it will be reported as-is.

nataliezeller1 replied:
@C0urante 
Invoking reporter::report doesn't cause the task to fail, it just writes to the dlq.
If the task should fail according to its configuration (or according to the error - for example IOException), it will fail after reporting to dlq (in case dlq is configured). 
If it's configured to ignore failures such as version conflicts, it will report to dlq and continue the task without failing.
For my use case it sounds reasonable...am I missing something?
There are some tests I've added to BulkProcessorTest that might clarify this behavior

C0urante replied:
@nataliezeller1 Invoking `reporter::report` actually can cause a task to fail; see more in [KIP-610](https://cwiki.apache.org/confluence/display/KAFKA/KIP-610%3A+Error+Reporting+in+Sink+Connectors#KIP610:ErrorReportinginSinkConnectors-ErrorsTolerance), which introduced this feature:

> The Errant Record Reporter will adhere to the existing DLQ error tolerance functionality. For example, if `errors.tolerance` is set to `ALL`, all errors will be tolerated; if the property is set to `NONE`, then the Errant Record Reporter will throw an exception detailing that the tolerance has been exceeded. Even if the developer chooses to catch and swallow any of these exceptions thrown during `report(...)`, the task is guaranteed to be killed following the completion of `put(...)`.

nataliezeller1 replied:
@C0urante Thanks, I see your point.
I've changed the logic to report to dlq only in these cases:
1. There was a failing error
2. errors.tolerance is ALL and an error occurred (including error types that are configured to ignore or warn)

I've also added a warning to alert on cases where the connector is configured with a dlq and error.tolerance=none, but behavior.on.version.conflict or behavior.on.malformed.documents are set to ignore/warn. 
In these cases, the record will not be reported to dlq in order to avoid failing the task.

C0urante replied:
@nataliezeller1 Thanks, glad we're on the same page.

I wonder if it's necessary to check the `errors.tolerance` property at all? I'm a little worried about that kind of logic in case the Kafka Connect framework is expanded and a third (or fourth, or fifth, etc.) possible value for that property is added.

I really think it'd be best to only invoke the reporter on cases where we would otherwise fail the connector, and not for issues that the user has explicitly configured the connector to ignore.

If users want to report everything to the DLQ, they could do this by setting `errors.tolerance` to `all` in the connector config, and setting the `behavior.on.<malforned.docs|version.conflict>` properties to `fail` (or leaving them unset, since `fail` is the default for them).

If we want to get really fancy, we could add a new `report` value for those properties that 1) Forces an assertion on task startup that an `ErrantRecordReporter` is available (and fails the task otherwise) and 2) causes failures related to each specific property to be passed to the reporter. This could be a little easier for new users to grok since it doesn't imply that the connector will "fail" on those errors, but it could also be overkill.

nataliezeller1 replied:
@C0urante 
In our use case, we want to move invalid events (ones that fail to be indexed due to incompatible schema) to dlq, without failing the task. 
Adding the "report" value can answer our needs, as long as it suggests that we'll report to dlq in any case of failure (without checking the errors.tolerance property), so the task might be failed in the reporter in case errors.tolerance is none (in our case it will be "all"). I can make this change if we are aligned on the solution now :-)

C0urante replied:
> Adding the "report" value can answer our needs, as long as it suggests that we'll report to dlq in any case of failure (without checking the errors.tolerance property)

Sounds good to me! Looking forward to the next patch; hoping we can get this in by the end of the week. Thanks for bearing with me on the design discussion!

nataliezeller1 replied:
Hi @C0urante 
My last commit contains the changes we've discussed - 
Now we'll report to DLQ only when an error related to 'behavior.on.version.conflict'/'behavior.on.malformed.documents' occurs and "report" is configured for the respective property. 
The connector will fail to start if "report" is configured for one of these properties but no DLQ topic is configured.
Sorry for the delay, hope it looks OK now :)

nataliezeller1 commented on src/main/java/io/aiven/kafka/connect/opensearch/BulkProcessor.java:
```diff
@@ -401,6 +421,25 @@ private BulkResponse execute() throws Exception {
+        private void reportMalformedRecord(final BulkItemResponse itemResponse) {
+            try {
+                if (reporter != null) {
+                    LOGGER.info("Reporting on record failure to DLQ");
```
Fixed

nataliezeller1 commented on src/main/java/io/aiven/kafka/connect/opensearch/BulkProcessor.java:
```diff
@@ -401,6 +421,25 @@ private BulkResponse execute() throws Exception {
+                        LOGGER.error("Only Map objects supported");
+                        return;
+                    }
+                    final String errorMsg = Optional.ofNullable(itemResponse.getFailureMessage())
```
Done

nataliezeller1 commented on src/main/java/io/aiven/kafka/connect/opensearch/BulkProcessor.java:
```diff
@@ -401,6 +421,25 @@ private BulkResponse execute() throws Exception {
+
+                }
+            } catch (final Exception e) {
+                LOGGER.error("An error occurred when reporting on record failure to DLQ: {}", e.getMessage(), e);
```
error message was refactored as suggested

nataliezeller1 commented:
Hi @C0urante ,
Thanks for the feedback!
I've addressed all your comments, please review my changes and let me know if there are more comments/questions.

C0urante commented on Jenkinsfile:
```diff
@@ -0,0 +1,37 @@
+@Library('ni-utils') _
```
Was this file included by mistake?

C0urante commented on src/main/java/io/aiven/kafka/connect/opensearch/OpensearchSinkConnectorConfig.java:
```diff
@@ -397,13 +404,38 @@ private static void addConversionConfigs(final ConfigDef configDef) {
                 Width.SHORT,
-                "Behavior on document's version conflict (optimistic locking)");
+                "Behavior on document's version conflict (optimistic locking)"
+        ).define(
+                ERRORS_TOLERANCE_CONFIG,
+                Type.STRING,
+                BulkProcessor.ErrorToleranceType.DEFAULT.toString(),
+                BulkProcessor.ErrorToleranceType.VALIDATOR,
+                Importance.LOW,
+                ERRORS_TOLERANCE_DOC,
+                group,
+                ++order,
+                Width.SHORT,
+                "Error Tolerance");
```
We should not be redefining these properties in the connector config; they should be left up to the Connect runtime.

C0urante commented on src/main/java/io/aiven/kafka/connect/opensearch/OpensearchSinkConnectorConfig.java:
```diff
@@ -397,13 +404,38 @@ private static void addConversionConfigs(final ConfigDef configDef) {
+
+    private void validateDeadLetterQueueConfig(final Map<String, String> props) {
+        final String dqlTopicStr = props.get(DLQ_TOPIC_NAME_CONFIG);
+        final boolean dlqIsConfigured = dqlTopicStr != null && !dqlTopicStr.isEmpty();
```
This doesn't cover all cases. The connector may be instantiated with a non-null `ErrantRecordReporter` even if it isn't configured to use a DLQ topic (see [here](https://github.com/apache/kafka/blob/6e7e2e08a982c6667476a932e2571f8ef4b70dc3/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java#L984-L986), [here](https://github.com/apache/kafka/blob/trunk/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/SinkConnectorConfig.java#L167-L170), and [here](https://github.com/apache/kafka/blob/trunk/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/ConnectorConfig.java#L262-L264)).

I don't think we should rely on parsing user-supplied configurations that are defined outside of the connector (i.e., in the Connect runtime) to validate whether a reporter should be required or not.

Instead, high-level, I'm thinking we might do this:
- Add a `requiresErrantRecordReporter` method (feel free to rename) to the `OpensearchSinkConnectorConfig` class that returns whether or not the connector requires an `ErrantRecordReporter` instance to function properly
- In the `OpensearchSinkTask` class, check whether an `ErrantRecordReporter` instance is required (using the method above); if it is, and if [`getErrantRecordReporter`](https://github.com/Natural-Intelligence/opensearch-connector-for-apache-kafka/blob/e23ff3c33e5fd5c5014ecae0a90de3d8372f5723/src/main/java/io/aiven/kafka/connect/opensearch/OpensearchSinkTask.java#L88) returns `null`, throw an exception explaining the problem to the user

This way, we don't have to redefine any runtime-defined properties (like `errors.tolerance`), and we cover all our bases for what may cause the connector to be instantiated with a non-null `ErrantRecordReporter`, including any future changes to that API like allowing developers to define their own pluggable error-handling logic.

Does that seem reasonable?

C0urante commented on src/main/java/io/aiven/kafka/connect/opensearch/BulkProcessor.java:
```diff
@@ -594,4 +670,53 @@ public String toString() {
+            private final ConfigDef.ValidString validator = ConfigDef.ValidString.in(names());
+
+            @Override
+            public void ensureValid(final String name, final Object value) {
+                if (value instanceof String) {
+                    final String lowerCaseStringValue = ((String) value).toLowerCase(Locale.ROOT);
+                    validator.ensureValid(name, lowerCaseStringValue);
+                } else {
+                    validator.ensureValid(name, value);
+                }
+            }
+
+            // Overridden here so that ConfigDef.toEnrichedRst shows possible values correctly
+            @Override
+            public String toString() {
+                return validator.toString();
+            }
+
+        };
+
+        public static String[] names() {
+            final ErrorToleranceType[] errorToleranceTypes = values();
+            final String[] result = new String[errorToleranceTypes.length];
+
+            for (int i = 0; i < errorToleranceTypes.length; i++) {
+                result[i] = errorToleranceTypes[i].toString();
+            }
+
+            return result;
+        }
+
+        public static ErrorToleranceType forValue(final String value) {
+            return valueOf(value.toUpperCase(Locale.ROOT));
+        }
+
+        @Override
+        public String toString() {
+            return name().toLowerCase(Locale.ROOT);
+        }
+    }
```
Similar to comments in the config class--we shouldn't be redefining this logic in the connector.
```suggestion
```

C0urante commented on src/main/java/io/aiven/kafka/connect/opensearch/BulkProcessor.java:
```diff
@@ -110,6 +125,18 @@ public BulkProcessor(final Time time,
                 BehaviorOnMalformedDoc.WARN);
         }
+
+        if (reporter != null && this.errorToleranceType == ErrorToleranceType.NONE
+                && (config.behaviorOnMalformedDoc() != BehaviorOnMalformedDoc.FAIL
+                || config.behaviorOnVersionConflict() != BehaviorOnVersionConflict.FAIL)) {
+            LOGGER.warn("The '{}' is set to `{}`, which means that only errors that would cause the task to fail"
+                        + " will be reported to the dead letter queue. You may consider setting property '{}' to '{}'"
+                        + " to deal with errors that are configured to be ignored or warned",
+                    OpensearchSinkConnectorConfig.ERRORS_TOLERANCE_CONFIG,
+                    ErrorToleranceType.NONE,
+                    OpensearchSinkConnectorConfig.ERRORS_TOLERANCE_CONFIG,
+                    ErrorToleranceType.ALL);
+        }
```
Is this still accurate? Don't we report errors (which does not necessarily mean send them to a DLQ topic!) if the behavior on malformed docs or on version conflicts is set to `REPORT` now?

I'm not sure we need any warning messages here--for all of the permutations I can think of for properties related to errors tolerance, behavior on malformed docs/version conflicts, etc., there's at least one circumstance in which each would make sense.

C0urante commented on src/main/java/io/aiven/kafka/connect/opensearch/BulkProcessor.java:
```diff
@@ -363,11 +391,19 @@ public BulkResponse call() throws Exception {
             }
         }
 
+        private void reportToDlq(final String errorMessage, final SinkRecord batchRecord) {
```
We really can't assume that this will end up in a DLQ topic :)
```suggestion
        private void sendToReporter(final String errorMessage, final SinkRecord batchRecord) {
```
(feel free to change the name--but nothing with DLQ in it!)

C0urante commented on src/main/java/io/aiven/kafka/connect/opensearch/BulkProcessor.java:
```diff
@@ -499,7 +573,8 @@ private static ConnectException toConnectException(final Throwable t) {
         WARN,
-        FAIL;
+        FAIL,
+        REPORT;
```
We should update the `OpensearchSinkConnectorConfig.BEHAVIOR_ON_MALFORMED_DOCS_DOC` docstring to explain what this new value will do.

C0urante commented on src/main/java/io/aiven/kafka/connect/opensearch/BulkProcessor.java:
```diff
@@ -545,11 +620,12 @@ public String toString() {
         WARN,
-        FAIL;
+        FAIL,
+        REPORT;
```
Same thought RE updating docstring in the `OpensearchConnectorConfig` class.

C0urante reviewed (commented):
Thanks @nataliezeller1! Looking a lot better.

nataliezeller1 commented:
Hi @C0urante
I've made the last changes you requested, would be great if you can review.
Thanks

C0urante commented on src/main/java/io/aiven/kafka/connect/opensearch/BulkProcessor.java:
```diff
@@ -410,6 +432,15 @@ private void handleVersionConflict(final BulkItemResponse bulkItemResponse) {
+                case REPORT:
+                    final String errorMessage =
+                            String.format("Encountered a version conflict when executing batch %s of %s"
+                                            + " records. Reporting to DLQ and will keep an existing record."
```
This will not necessarily cause anything to be sent to a DLQ topic, so we shouldn't assume that in our log message:

```suggestion
                                            + " records. Reporting this error to the Connect runtime and will keep an existing record."
```

C0urante commented on src/main/java/io/aiven/kafka/connect/opensearch/BulkProcessor.java:
```diff
@@ -439,6 +470,15 @@ private void handleMalformedDoc(final BulkItemResponse bulkItemResponse) {
+                case REPORT:
+                    final String errorMessage =
+                            String.format("Encountered a version conflict when executing batch %s of %s"
+                                            + " records. Reporting to DLQ and will not index record."
```
Same thought RE assuming a DLQ topic:
```suggestion
                                            + " records. Reporting this error to the Connect runtime and will not index record."
```

C0urante commented on src/main/java/io/aiven/kafka/connect/opensearch/OpensearchSinkConnectorConfig.java:
```diff
@@ -145,13 +145,18 @@ public class OpensearchSinkConnectorConfig extends AbstractConfig {
+            + "'warn' - log a warning message and keep the existing record, 'report' - report to errant record reporter"
+            + " and keep the existing record, 'fail' - fail the task.";
+
+    public static final String DLQ_TOPIC_NAME_CONFIG =  "errors.deadletterqueue.topic.name";
```
I think we can just remove this field completely, since it's not used anywhere in the main codebase, and is not actually necessary for testing (see comments in `BulkProcessorTest`).

C0urante commented on src/test/java/io/aiven/kafka/connect/opensearch/OpensearchSinkTaskTest.java:
```diff
@@ -51,4 +70,22 @@ void convertTopicToIndexName() {
     }
 
+    @Test
+    void failToStartWhenReportIsConfiguredAndErrantRecordReporterIsMissing(final @Mock SinkTaskContext context) {
```
Whoa! Didn't realize you could use `@Mock` objects as method parameters for test cases. Super cool!

C0urante commented on src/test/java/io/aiven/kafka/connect/opensearch/BulkProcessorTest.java:
```diff
@@ -486,13 +494,140 @@ public void ignoreOnVersionConfict(final @Mock RestHighLevelClient client) throw
+                LINGER_MS_CONFIG, "1000",
+                MAX_RETRIES_CONFIG, "3",
+                READ_TIMEOUT_MS_CONFIG, "1",
+                DLQ_TOPIC_NAME_CONFIG, "dlq_topic",
```
This line can be removed and the tests will still pass. IMO it's better to remove it since it better illustrates that the only thing that the `BulkProcessor` cares about is whether it's configured with the `report` value for handling version conflicts or malformed docs, and doesn't care about framework-level properties like the DLQ topic name or errors tolerance setting.

C0urante commented on src/test/java/io/aiven/kafka/connect/opensearch/BulkProcessorTest.java:
```diff
@@ -486,13 +494,140 @@ public void ignoreOnVersionConfict(final @Mock RestHighLevelClient client) throw
+                LINGER_MS_CONFIG, "1000",
+                MAX_RETRIES_CONFIG, "3",
+                READ_TIMEOUT_MS_CONFIG, "1",
+                DLQ_TOPIC_NAME_CONFIG, "dlq_topic",
```
Same thought as above; this line can and should be removed.

C0urante commented on src/test/java/io/aiven/kafka/connect/opensearch/OpensearchSinkTaskTest.java:
```diff
@@ -51,4 +70,22 @@ void convertTopicToIndexName() {
+                BEHAVIOR_ON_VERSION_CONFLICT_CONFIG, BulkProcessor.BehaviorOnMalformedDoc.REPORT.toString()
+        );
+        final var task = new OpensearchSinkTask();
+        task.initialize(context);
```
Should we explicitly set up an expectation for `context::errantRecordReporter` to return `null`, to make the intent of this test clearer?

Also, can we add a case where `context::errantRecordReporter` throws a `NoSuchMethodError` and handles it gracefully (i.e., in the same way it does in this test case)?

C0urante reviewed (commented):
Thanks @nataliezeller1, this is really close! I think once all the comments here are addressed this will be ready to merge.

nataliezeller1 commented:
Hi @C0urante ,
My last commit includes the changes you've requested

C0urante reviewed (approved):
Thanks so much @nataliezeller1, LGTM!

Would you mind squashing this PR so that it consists of a single commit? Once that's done, I can ping one of my colleagues who has write access on this repo and we can hopefully get it merged.

C0urante commented:
Oh, and can you rebase on the latest `main` to fix the merge conflicts?

C0urante reviewed (approved):
LGTM, thanks @nataliezeller1!

C0urante commented:
@ivanyu can you take a look at this when you have a moment?

C0urante commented:
Thanks again @nataliezeller1, really appreciate all your hard work on this valuable new feature!

reta commented on src/main/java/io/aiven/kafka/connect/opensearch/OpensearchSinkConnectorConfig.java:
```diff
@@ -157,13 +157,16 @@ public class OpensearchSinkConnectorConfig extends AbstractConfig {
             + "OpenSearch rejects due to some malformation of the document itself, such as an index"
             + " mapping conflict or a field name containing illegal characters. Valid options are "
-            + "``ignore``, ``warn``, and ``fail``.";
+            + "``ignore`` - do not index the record, ``warn`` - log a warning message and do not index the record, "
```
@C0urante could you please update docs [1] with these configuration changes? cc @willyborankin 

[1] https://github.com/aiven/opensearch-connector-for-apache-kafka/blob/main/docs/opensearch-sink-connector-config-options.rst
